### PR TITLE
Documentation for graph classes

### PR DESCRIPTION
**Pull request by mscroggs**
- Added a documentation page aimed at users
- Added a documentation page aimed at developers
- Added some missing documentation of function args (see #183)
- Made the type hints of some functions consistently use `jax.Array` instead of various `np.typing` things (see #155)

~This documentation assumes that #189 will be merged and will need editing if we decide to not merge that.~ #189 was merged

### CODE REVIEW

File: docs/developers/graph.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Graphs in causalprog
+
+In causalprog, directed acyclic graphs (DAGs) are used to represent causal problems.
+This documentation page describes how these graphs are implemented, and is aimed
+at developers of the library. Documentation for users can be found in [the documentation
```
Comment (willGraham01) [inline docs/developers/graph.md]: I won't do this everywhere, but it's generally good to start each new sentence in a markdown file on a new line. That way when the linter reports an error in your prose, you know the exact sentence rather than just the paragraph in which the error was detected (similar ethos to LaTeX).

File: docs/developers/graph.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+for users](../users/graph.md).
+
+## Graphs
+
+Graphs in causalprog are internally stored as [networkx](https://networkx.org/) graph, with nodes
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
Graphs in causalprog are internally stored as [networkx](https://networkx.org/) graphs, with nodes
```

File: docs/developers/graph.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+being instances of subclasses of `causalprog.graph.base.Node`. The interface to networkx is hidden
+from users, with methods definite in the node and graph classes making the direct calls to
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
from users, with methods defined in the node and graph classes making the direct calls to
```

or am I misunderstanding the use of `definite` here?

File: docs/developers/graph.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+networkx. This should make is easier to replace networkx with another graph library in future
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
networkx. This should make it easier to replace networkx with another graph library in future
```

also surprised you didn't take the opportunity to mention `rustworkx` here 😁

File: docs/developers/graph.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+All graph and node classes in causalprog inherit from `causalprog._abc.labelled.Labelled` which
+enforces that each instance has a label set at the point of initialisation.
+
+## Nodes
+
+All graph nodes in causalprog must inheret from the `causalprog.graph.base.Node` base class. This
+class has the following abstract methods that must be implemented:
+
+- `evaluate` returns an evaluation of the node given valuse of its parents.
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
- `evaluate` returns an evaluation of the node given values of its parents.
```

File: docs/developers/graph.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+- `copy` makes a (deep) copy of the node.
+- `parents` returns a the node's parents. `parents` is a property instead of a method.
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
- `parents` returns the node's parents.
  `parents` is a property instead of a method.
```

File: docs/developers/graph.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+- `sample` samples a value from the node. This function is no longer used in the examples and could
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
- `sample` samples value(s) from the node.
  This function is no longer used in the examples and could
```

File: docs/developers/graph.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  be considered for removal.
+
+Inside the initialiser of any subclass of `causalprog.graph.base.Node`, the `super()` initialiser
+function must be called, with `label` given as a required keyword argument and `shape` as an
+optional second keyword argument defauilting to `()` for a scalar.
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
optional second keyword argument, defaulting to `()` for a scalar.
```

File: docs/developers/graph.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+## Algorithms
+
+In general, functions that act on a single node or return information about the full graph are
+implemented as methods of properties of the graph or node classes, while functions that iterate
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
implemented as methods or properties of the graph or node classes, while functions that iterate
```

File: docs/developers/graph.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+through all nodes in a graph, copy and modify graph, or are more computationally involved are
```
Comment (willGraham01) [inline docs/developers/graph.md]:
```suggestion
through all nodes in a graph, copy and / or modify it, or are more computationally involved are
```

File: docs/developers/graph.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+implemented as functions in `causalprog.algorithms`.
+
```
Comment (willGraham01) [inline docs/developers/graph.md:57]: It might be worth a paragraph here noting that future development should focus on emphasizing an `algorithms`-focused approach. Right now our various builder functions directly call the `compute` methods of the nodes, but in a perfect world we would be doing the proper `algorithm.nodes_down_to_outcome` calls instead, to not depend on the specific graph layout.

File: docs/users/graph.md [hunk 12]
```diff
@@ -0,0 +1,300 @@
+# Creating a using graphs in causalprog
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
# Creating and using graphs in causalprog
```

File: docs/users/graph.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+from causalprog.graph import ConstantNode
+
+one = ConstantNode(label="one", value=1.0)
+vector = ConstantNode(label="v", value=jnp.array([1.0, 1.0, 2.0]))
+```
+
+### `DataNode`
+
+A `DataNode` is a node that represents a constant value that is not known
+when the node is created. These nodes has one required keyword arguments
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
when the node is created. These nodes have one required keyword argument
```

File: docs/users/graph.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+from causalprog.graph import DataNode
+
+scalar = DataNode(label="my_scalar")
+vector = DataNode(label="my_vector", shape=(5, ))
+matrix = DataNode(label="my_matrix", shape=(4, 2))
+```
+
+### Random variable nodes
+
+Random variable node represent random variables (RVs) in a causal problem.
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
Random variable nodes represent random variables (RVs) in a causal problem.
```

File: docs/users/graph.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+There are two types of random variable node in causalprog:
+`ContinuousRandomVariableNode` and `DiscreteRandomVariableNode`. Both of
+these must be passed the a `label` for the node as a required keyword
+argument, and can take a number of additional keyword arguments: the `shape`
+of the data that the RV outputs, a function to `compute` the value of the RV
+from the values of its parents, and a list of `parents` of the RV node.
+Discrete RV nodes must be passed an addition required keyword argument: a
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
Discrete RV nodes must be passed an additional required keyword argument: a
```

File: docs/users/graph.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+list of possible `values` that the RV can output.
+
+The `compute` function should take a single input, and will be passes a
+dictionary containing the values of the parents of the node, with node
+labels as keys. It should return the value of the RV.
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
The `compute` function should take two inputs; 

- The first being a dictionary containing the values of the parents of the node, with the keys corresponding to the parent node labels.
- The second being a dictionary containing the current values of the parameters that parametrize the `compute` function.

Typically, the `compute` function will be some kind of predictive model or neural network that predicts the value of the RV from the parent node values (first input) and the parameters - in this case the weights and biases - of the neural network (second input).
The [`FunctionalMLP` class](FIXME-link-to-Sams-docs) demonstrates one such set of `compute` functions.

`compute` should return the value of the RV, as computed / predicted using the inputs.
```

`compute` should be an `MLPAlias` right? So it should take 2 input arguments. Can include a X-ref here to Sam's docs on MLPs when ready so that you don't have to provide a concrete example here, and can even X-ref to the example in #187 where we use deterministic functions.

File: docs/users/graph.md [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+
+```python
+from causalprog.graph import ContinuousRandomVariableNode, DiscreteRandomVariableNode
+
+node1 = DiscreteRandomVariableNode(values=[1.0, 1.5, 2.0], label="X")
+node2 = ContinuousRandomVariableNode(label="Y")
+node3 = ContinuousRandomVariableNode(
+    label="2Y",
+    compute=lambda values: values["Y"] * 2,
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
    compute=lambda values, parameters: values["Y"] * parameters["scale"],
```

To be consistent with the notation above.

File: docs/users/graph.md [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+)
+```
+
+### `DistributionNode`
+
+Distrubution nodes represent the values of random variables (RVs) that can
+be sampled from. These nodes formed part of an earlier experimental version
+of the library and are not used in the current demonstration applications.
+
+## Ricardo's graph
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
## Continuous Treatment models
```

Just to move away from our "ricardo" terminology everywhere 😅

File: docs/users/graph.md [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+
+Many of the examples in causalprog use an example graph for a problem
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
Many of the examples in causalprog use an example graph, representing a continuous treatment model, for a problem
```

File: docs/users/graph.md [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+proposed by Ricardo Silva:
+
+![Illustration of the continuous treatment model that we discuss.](../diagrams/continuous-treatment-model.svg)
+
+causalprog provides a helper function to quickly generate this graph. This
+function must be passed five required keyword arguments that tell the graph
+how to compute the nodes `"u_x"`, `"u_y"`, `"phi_x"`, `"x"` and `"y"` from
+their parents.
```
Comment (willGraham01) [inline docs/users/graph.md:169]: May need to change this subject to how we resolve #156

File: docs/users/graph.md [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+
+```python
+from causalprog.graph.ricardo import example_model
+
+graph = example_model(
+    compute_u_x=lambda values: values["c"][0] + 1.0,
+    compute_u_y=lambda values: values["c"][1] * 2,
+    compute_phi_x=lambda values: values["l"],
+    compute_x=lambda values: values["z"] + values["phi_x"] - values["u_x"],
+    compute_y=lambda values: values["x"] * values["u_y"],
```
Comment (willGraham01) [inline docs/users/graph.md]: As above, these need to take 2 inputs. Though using `lambda values, _:` in each would solve this if you wanted a fast out.

Better would be to have them all of the form `lambda values, theta_{node_name}:` to be consistent with our examples.

File: docs/users/graph.md [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+)
+```
+
+## Using a graph
+
+This section of the documentation demonstrated how graphs created using
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
This section of the documentation demonstrates how graphs created using
```

File: docs/users/graph.md [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+successors of the node are the node's children, plus the children's
+children, and so on. In `causalprog` dictionaries mapping each node onto
+tuples of this predecessors and successors can be obtained using the
+properties `graph.predecessors` and `graph.successors`.
+
+### Ordered nodes
+
+The property `graph.ordered_nodes` and the method
+`graph.roots_down_to_outcome` can be used to obtain tuples of nodes ordered
+so that every node's parents appear before that node in the tuple. This
```
Comment (willGraham01) [inline docs/users/graph.md]:
```suggestion
so that; for every node $X$ in the graph, the parents of node $X$ appear before node $X$ in the returned tuple. This
```

Don't know if this is any clearer than what you already have though

File: docs/users/graph.md [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+```python
+import jax.numpy as jnp
+from causalprog.algorithms import evaluate
+
+value = evaluate(
+    graph,
+    "x",
+    {"l": jnp.array([2.0]), "z": jnp.array([1.0]), "c": 1.0},
+)
+```
```
Comment (willGraham01) [inline docs/users/graph.md:291]: Just realised that these methods will need to take a fourth argument, which is a dictionary whose keys are the node labels and whose values are the parameter values of that node. Again because of the 2-input requirement of the `compute` functions.

**Review by willGraham01 — CHANGES_REQUESTED** [review]
Besides a few typos, only major thing is that `compute` is geared up to be a function of 2 inputs to be consistent with `FunctionalMLP`.

The `evaluate` and `evaluate_down_to` algorithms in particular might need an adjustment because of this. But might want to save that for another PR, to keep this one limited to the `docs`.

File: docs/users/graph.md (hunk 21 above)
Comment (mscroggs) [reply docs/users/graph.md]: I'm going to update this in the next PR, as a few tests will need a matching update and it might make a mess.

Comment (mscroggs) [reply docs/users/graph.md]: Issue for this: #196

File: docs/users/graph.md (hunk 24 above)
Comment (mscroggs) [reply docs/users/graph.md:291]: Will update this as part of #196

File: docs/users/graph.md (hunk 16 above)
Comment (mscroggs) [reply docs/users/graph.md]: Opened #197 to remind us to add the link suggested here

**Review by willGraham01 — APPROVED** [review]
